skimage/future/graph/graph_merge.py

```diff
@@ -128,10 +134,16 @@ def merge_hierarchical(labels, rag, thresh, rag_copy, in_place_merge,
             merge_func(rag, src, dst)
             new_id = rag.merge_nodes(src, dst, weight_func)
             _revalidate_node_edges(rag, new_id, edge_heap)
+            if merge_trace:
+                label_map = np.arange(labels.max() + 1)
+                for ix, (n, d) in enumerate(rag.nodes_iter(data=True)):
+                    for label in d['labels']:
+                        label_map[label] = ix
+                trace.append(label_map[labels])
```

Comment:
Agreed.   For my specific use case, Vighnesh provide the following, which seems similar to your ideas in para 2 and 3:
```python
label_map = np.arange(labels.max() + 1)
label_map[:] = 2 # Label the rest of the graph as one region

# Label src as one region
for l in rag.node[src]['labels']:
    label_map[l] = 1 

for l in rag.node[dst]['labels']:
    label_map[l] = 1 

seg_list.append(labels[label_map])
```
A *merge tree* sounds closer to what I want.  I will have a look at viridis library, tree.py for some inspiration.

Thanks.